proxy/src/coco/git_helper.rs

```diff
@@ -0,0 +1,46 @@
+//! git-remote-rad git helper related functionality.
+
+use crate::config;
+
+use crate::error;
+use std::fs;
+use std::os::unix::fs::PermissionsExt;
+
+/// Filename of the git helper binary.
+pub const HELPER_BINARY_NAME: &str = "git-remote-rad";
+
+/// Check if a path contains an executable file.
+fn is_executable(path: &std::path::PathBuf) -> Result<bool, error::Error> {
+    let metadata = path.metadata()?;
+    let permissions = metadata.permissions();
+    Ok(permissions.mode() & 0o111 != 0)
+}
+
+/// Checks if the git-remote-rad helper is in a stable location and has the
+/// executable flag, if not copies the executable to the right place.
+///
+/// # Errors
+///
+///   * Could not get the path to directory where helper binaries should be stored.
+///   * Could not get the current working directory.
+///   * Could not create the path to binary directory.
+///   * Could not copy helper executable to the binary directory.
+pub fn setup() -> Result<(), error::Error> {
+    log::info!("Making sure git-remote-rad helper is set up");
+
+    let bin_dir = config::bin_dir()?;
+    let full_dest_path = bin_dir.join(HELPER_BINARY_NAME);
+
+    if full_dest_path.exists() && is_executable(&full_dest_path)? {
+        log::debug!("Git helper already exists at: {:?}", full_dest_path);
+        return Ok(());
+    }
+
+    let proxy_path = config::proxy_path()?;
+
+    fs::create_dir_all(bin_dir)?;
```

Comment:
So what happens if it exists and is not executable? I imagine this will blow up.

Reply:
It works just fine:

"Recursively create a directory and all of its parent components **if they are missing**."
https://doc.rust-lang.org/std/fs/fn.create_dir_all.html

Reply:
Ah true, it won't blow up. But we don't actually have an executable :)

Reply:
Actually, I'd also think that the `copy` below would fail. Because if you `copy foo/existing_file bar/existing_file` it'll break, no?

Reply:
I tested these cases manually and they all pass:
- no path exists
- no binary exists at path
- binary not executable at path

But I probably should just write tests for these...

Reply:
> binary not executable at path

How did you test this? Like would it mean that this code passes but we can't actually use the remote helper? :thinking: 

> But I probably should just write tests for these...

That'd be good :+1: You can add:
```rust
#[cfg(test)]
mod tests {
}
```
in this file :)

Reply:
> How did you test this?

Manually changed permissions of the helper:
```
» cd ~/.radicle/bin
» ls -la git-remote-rad
-rwxr-xr-x  1 rudolfs  staff  8731832 Aug 17 10:29 git-remote-rad
» chmod -x git-remote-rad
-rw-r--r--  1 rudolfs  staff  8731832 Aug 17 10:29 git-remote-rad
```

then started the app via `yarn start`:

```
» RUST_LOG=proxy=debug yarn start                                                                                                                                                                                           INFO  proxy::coco::git_helper > Making sure git-remote-rad helper is set up
DEBUG proxy::coco::git_helper > Copied git-remote-rad helper to: "/Users/rudolfs/.radicle/bin/git-remote-rad"
```

and checked that the remote helper is executable after the app had started:
```
» ls -la git-remote-rad
-rwxr-xr-x  1 rudolfs  staff  8731832 Aug 17 10:29 git-remote-rad
```

> Like would it mean that this code passes but we can't actually use the remote helper? 🤔

It means the git remote helper is in place and working as intended.

Reply:
Ah cool :+1: Thanks for the details :)

Reply:
https://github.com/radicle-dev/radicle-upstream/pull/769/commits/3fa334ad3f38b15e86e8433d6794c4bea4301b1d